### PR TITLE
Fix bug in Einsum implementation

### PR DESCRIPTION
**Pull request by hariharans29**
**Description**: Fix bug in the Einsum implementation for inputs that had at-least one input which had multi-axes reduction prior to the MatMul step. In the existing implementation, we dis-regard any earlier reductions in case there were multi-axes reductions involved.  

**Motivation and Context**
Resolve https://github.com/microsoft/onnxruntime/issues/7375

### CODE REVIEW

File: onnxruntime/test/providers/cpu/math/einsum_test.cc [hunk 1]
```diff
@@ -254,6 +254,16 @@ TEST(Einsum, ExplicitEinsumAsMatmul_2) {
   test.Run();
 }
 
+TEST(Einsum, ExplicitEinsumAsMatmul_3) {
+  OpTester test("Einsum", 12, onnxruntime::kOnnxDomain);
+  test.AddAttribute<std::string>("equation", "abc,cd,def->abd");
```
Comment (hariharans29) [inline onnxruntime/test/providers/cpu/math/einsum_test.cc:259]: In this test case, tensor "z" involves a multi-axes reduction in axes "e" and "f"

Comment (xadupre) [reply onnxruntime/test/providers/cpu/math/einsum_test.cc:259]: It works for right now, does it work also for left?

File: onnxruntime/test/providers/cpu/math/einsum_test.cc [hunk 2]
```diff
@@ -254,6 +254,16 @@ TEST(Einsum, ExplicitEinsumAsMatmul_2) {
   test.Run();
 }
 
+TEST(Einsum, ExplicitEinsumAsMatmul_3) {
+  OpTester test("Einsum", 12, onnxruntime::kOnnxDomain);
```
Comment (skottmckay) [inline onnxruntime/test/providers/cpu/math/einsum_test.cc:258]: There were a lot of test cases in this WIP PR: [https://github.com/microsoft/onnxruntime/pull/7374/files](https://github.com/microsoft/onnxruntime/pull/7374/files)

Could/should we include at least some of those?

File: onnxruntime/test/providers/cpu/math/einsum_test.cc (hunk 1 above)
Comment (hariharans29) [reply onnxruntime/test/providers/cpu/math/einsum_test.cc:259]: yes. To elaborate:

1) While processing pair-wise operands, if the left operand had multi-axes reduction, the code handles it now (the bug existed just like in the right operand's case) - I can add a test for this

2) If the very first operand (the very first operand is special-cased) had multi-axes reduction, the code handles that correctly

Comment (hariharans29) [reply onnxruntime/test/providers/cpu/math/einsum_test.cc:259]: Actually the tests included in 7374 (thanks!) are quite comprehensive and I just added it to this PR - there are no extra tests needed.

File: onnxruntime/test/providers/cpu/math/einsum_test.cc (hunk 2 above)
Comment (hariharans29) [reply onnxruntime/test/providers/cpu/math/einsum_test.cc:258]: Added all of them

**Review by skottmckay — APPROVED** [review]
:shipit:

**Pull request merged**